File: sdk/c/src/providers/maildrop.c
/**
 * Maildrop — https://maildrop.cx
 */
#include "tempmail_internal.h"
#include <curl/curl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#ifdef _WIN32
#define MD_STRICMP _stricmp
#else
#include <strings.h>
#define MD_STRICMP strcasecmp
#endif

#define MD_BASE "https://maildrop.cx"
#define MD_EXCLUDED "transformer.edu.kg"

static const char *md_headers[] = {
    "Accept: application/json, text/plain, */*",
    "Accept-Language: zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
    "Cache-Control: no-cache",
    "DNT: 1",
    "Pragma: no-cache",
    "Referer: https://maildrop.cx/zh-cn/app",
    "sec-ch-ua: \"Chromium\";v=\"146\", \"Not-A.Brand\";v=\"24\", \"Microsoft "
    "Edge\";v=\"146\"",
    "sec-ch-ua-mobile: ?0",
    "sec-ch-ua-platform: \"Windows\"",
    "sec-fetch-dest: empty",
    "sec-fetch-mode: cors",
    "sec-fetch-site: same-origin",
    "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
    "x-requested-with: XMLHttpRequest",
    NULL};

static void md_random_local(char *buf, int len) {
  const char chars[] = "abcdefghijklmnopqrstuvwxyz0123456789";
  srand((unsigned)time(NULL));
  for (int i = 0; i < len; i++)
    buf[i] = chars[rand() % (int)(sizeof(chars) - 1)];
  buf[len] = '\0';
}

static char *md_curl_escape(const char *s) {
  if (!s)
    return tm_strdup("");
  CURL *c = curl_easy_init();
  if (!c)
    return tm_strdup(s);
  char *e = curl_easy_escape(c, s, 0);
  curl_easy_cleanup(c);
  if (!e)
    return tm_strdup(s);
  char *d = tm_strdup(e);
  curl_free(e);
  return d;
}

static void md_pick_domain(char domains[][256], int nd, const char *preferred,
                           char *out, size_t cap) {
  if (!out || cap < 2)
    return;
  out[0] = '\0';
  if (nd <= 0)
    return;
  if (preferred && preferred[0]) {
    for (int i = 0; i < nd; i++) {
      if (MD_STRICMP(domains[i], preferred) == 0) {
        strncpy(out, domains[i], cap - 1);
        out[cap - 1] = '\0';
        return;
      }
    }
    return;
  }
  int pick = rand() % nd;
  strncpy(out, domains[pick], cap - 1);
  out[cap - 1] = '\0';
}

tm_email_info_t *tm_provider_maildrop_generate(const char *domain) {
  srand((unsigned)time(NULL));
  tm_http_response_t *resp = tm_http_request(
      TM_HTTP_GET, MD_BASE "/api/suffixes.php", md_headers, NULL, 15);
  if (!resp || resp->status < 200 || resp->status >= 300) {
    tm_http_response_free(resp);
    return NULL;
  }

  cJSON *root = cJSON_Parse(resp->body);
  tm_http_response_free(resp);
  if (!root || !cJSON_IsArray(root)) {
    cJSON_Delete(root);
    return NULL;
  }

  char domains[64][256];
  int nd = 0;
  int n = cJSON_GetArraySize(root);
  for (int i = 0; i < n && nd < 64; i++) {
    cJSON *it = cJSON_GetArrayItem(root, i);
    const char *s = cJSON_GetStringValue(it);
    if (s && s[0] && MD_STRICMP(s, MD_EXCLUDED) != 0) {
      strncpy(domains[nd], s, sizeof(domains[0]) - 1);
      domains[nd][sizeof(domains[0]) - 1] = '\0';
      nd++;
    }
  }
  cJSON_Delete(root);
  if (nd == 0)
    return NULL;

  char pickdom[256];
  md_pick_domain(domains, nd, domain, pickdom, sizeof(pickdom));
  if (!pickdom[0])
    return NULL;

  char local[16];
  md_random_local(local, 10);

  char email[320];
  snprintf(email, sizeof(email), "%s@%s", local, pickdom);

  tm_email_info_t *info = tm_email_info_new();
  info->channel = CHANNEL_MAILDROP;
  info->email = tm_strdup(email);
  info->token = tm_strdup(email);
  return info;
}

/**
 * 通过详情接口获取单封邮件完整内容
 * GET /api/email_content.php?id={id}
 * 详情响应字段（从前端代码确认）:
 *   - content: 完整 HTML 正文
 *   - subject / from_addr / date: 邮件元数据
 *   - attachment: JSON 字符串数组 [{filename, path, size}]（可能为空）
 * 返回解析后的 cJSON 对象（调用方负责释放），失败返回 NULL
 */
static cJSON *md_fetch_detail(const char *id) {
  if (!id || !id[0]) return NULL;
  char *esc = md_curl_escape(id);
  char url[800];
  snprintf(url, sizeof(url), "%s/api/email_content.php?id=%s", MD_BASE, esc);
  free(esc);
  tm_http_response_t *resp =
      tm_http_request(TM_HTTP_GET, url, md_headers, NULL, 15);
  if (!resp || resp->status < 200 || resp->status >= 300) {
    tm_http_response_free(resp);
    return NULL;
  }
  cJSON *detail = cJSON_Parse(resp->body);
  tm_http_response_free(resp);
  if (!detail || !cJSON_IsObject(detail)) {
    cJSON_Delete(detail);
    return NULL;
  }
  return detail;
}

/**
 * 将详情接口返回的 content 字段映射为 html 字段
 * 详情的其他字段（subject/from_addr/date）优先级更高
 */
static void md_merge_detail_into_raw(cJSON *raw, cJSON *detail) {
  if (!raw || !detail) return;
  /* content -> html */
  cJSON *content = cJSON_GetObjectItemCaseSensitive(detail, "content");
  if (cJSON_IsString(content) && content->valuestring && content->valuestring[0]) {
    cJSON_DeleteItemFromObjectCaseSensitive(raw, "html");
    cJSON_AddStringToObject(raw, "html", content->valuestring);
  }
  /* from_addr / subject / date 使用详情覆盖 */
  for (int i = 0; i < 3; i++) {
    const char *keys[] = {"from_addr", "subject", "date"};
    cJSON *item = cJSON_GetObjectItemCaseSensitive(detail, keys[i]);
    if (cJSON_IsString(item) && item->valuestring && item->valuestring[0]) {
      cJSON_DeleteItemFromObjectCaseSensitive(raw, keys[i]);
      cJSON_AddStringToObject(raw, keys[i], item->valuestring);
    }
  }
  /* attachment（JSON字符串）解析为附件数组 */
  cJSON *att = cJSON_GetObjectItemCaseSensitive(detail, "attachment");
  if (cJSON_IsString(att) && att->valuestring && att->valuestring[0]) {
    cJSON *att_arr = cJSON_Parse(att->valuestring);
    if (att_arr && cJSON_IsArray(att_arr)) {
      cJSON_DeleteItemFromObjectCaseSensitive(raw, "attachments");
      cJSON_AddItemToObject(raw, "attachments", att_arr);
    } else {
      cJSON_Delete(att_arr);
    }
  }
}

tm_email_t *tm_provider_maildrop_get_emails(const char *token,
                                            const char *email, int *count) {
  *count = -1;
  const char *addr = (email && email[0]) ? email : token;
  if (!addr || !addr[0])
    return NULL;

  char *esc = md_curl_escape(addr);
  char url[800];
  snprintf(url, sizeof(url), "%s/api/emails.php?addr=%s&page=1&limit=20",
           MD_BASE, esc);
  free(esc);

  tm_http_response_t *resp =
      tm_http_request(TM_HTTP_GET, url, md_headers, NULL, 15);
  if (!resp || resp->status < 200 || resp->status >= 300) {
    tm_http_response_free(resp);
    return NULL;
  }

  cJSON *root = cJSON_Parse(resp->body);
  tm_http_response_free(resp);
  if (!root)
    return NULL;

  cJSON *arr = cJSON_GetObjectItemCaseSensitive(root, "emails");
  int n = (arr && cJSON_IsArray(arr)) ? cJSON_GetArraySize(arr) : 0;
  *count = n;
  if (n == 0) {
    cJSON_Delete(root);
    return NULL;
  }

  tm_email_t *emails = tm_emails_new(n);
  if (!emails) {
    cJSON_Delete(root);
    *count = -1;
    return NULL;
  }

  for (int i = 0; i < n; i++) {
    cJSON *raw = cJSON_GetArrayItem(arr, i);

    /* 提取邮件 ID 用于详情二拉 */
    char id_buf[64] = {0};
    cJSON *id_item = cJSON_GetObjectItemCaseSensitive(raw, "id");
    if (cJSON_IsString(id_item) && id_item->valuestring) {
      strncpy(id_buf, id_item->valuestring, sizeof(id_buf) - 1);
    } else if (cJSON_IsNumber(id_item)) {
      snprintf(id_buf, sizeof(id_buf), "%.0f", id_item->valuedouble);
    }

    /* 无条件调用详情接口，用 content/from_addr/subject/date/attachment 覆盖列表字段 */
    if (id_buf[0]) {
      cJSON *detail = md_fetch_detail(id_buf);
      if (detail) {
        md_merge_detail_into_raw(raw, detail);
        cJSON_Delete(detail);
      }
    }

    emails[i] = tm_normalize_email(raw, addr);
  }
  cJSON_Delete(root);
  return emails;
}

File: sdk/rust/src/providers/maildrop.rs
use crate::config::{block_on, get_current_ua, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};
use rand::Rng;
use serde_json::Value;

const BASE: &str = "https://maildrop.cx";
const EXCLUDED_SUFFIX: &str = "transformer.edu.kg";

fn md_headers(b: wreq::RequestBuilder) -> wreq::RequestBuilder {
    b.header("Accept", "application/json, text/plain, */*")
        .header(
            "Accept-Language",
            "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
        )
        .header("Cache-Control", "no-cache")
        .header("DNT", "1")
        .header("Pragma", "no-cache")
        .header("Referer", "https://maildrop.cx/zh-cn/app")
        .header(
            "sec-ch-ua",
            r#""Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146""#,
        )
        .header("sec-ch-ua-mobile", "?0")
        .header("sec-ch-ua-platform", r#""Windows""#)
        .header("sec-fetch-dest", "empty")
        .header("sec-fetch-mode", "cors")
        .header("sec-fetch-site", "same-origin")
        .header("User-Agent", get_current_ua())
        .header("x-requested-with", "XMLHttpRequest")
}

fn random_local(len: usize) -> String {
    let chars: Vec<char> = "abcdefghijklmnopqrstuvwxyz0123456789".chars().collect();
    let mut rng = rand::thread_rng();
    (0..len)
        .map(|_| chars[rng.gen_range(0..chars.len())])
        .collect()
}

fn fetch_suffixes() -> Result<Vec<String>, String> {
    block_on(async {
        let resp = md_headers(http_client().get(format!("{}/api/suffixes.php", BASE)))
            .send()
            .await
            .map_err(|e| e.to_string())?;
        if !resp.status().is_success() {
            return Err(format!("maildrop suffixes {}", resp.status()));
        }
        let v: Value = resp.json().await.map_err(|e| e.to_string())?;
        let arr = v
            .as_array()
            .ok_or_else(|| "maildrop: invalid suffixes".to_string())?;
        let ex = EXCLUDED_SUFFIX.to_lowercase();
        let mut out = Vec::new();
        for x in arr {
            if let Some(s) = x.as_str() {
                let t = s.trim();
                if !t.is_empty() && t.to_lowercase() != ex {
                    out.push(t.to_string());
                }
            }
        }
        if out.is_empty() {
            return Err("maildrop: no domains".into());
        }
        Ok(out)
    })
}

fn pick_suffix(suffixes: &[String], preferred: Option<&str>) -> Result<String, String> {
    if let Some(p) = preferred {
        let pl = p.trim().to_lowercase();
        if !pl.is_empty() {
            if let Some(hit) = suffixes.iter().find(|d| d.to_lowercase() == pl) {
                return Ok(hit.clone());
            }
            return Err(format!("maildrop: domain not available: {}", pl));
        }
    }
    let mut rng = rand::thread_rng();
    Ok(suffixes[rng.gen_range(0..suffixes.len())].clone())
}

pub fn generate_email(domain: Option<&str>) -> Result<EmailInfo, String> {
    let suffixes = fetch_suffixes()?;
    let dom = pick_suffix(&suffixes, domain)?;
    let local = random_local(10);
    let email = format!("{}@{}", local, dom);
    Ok(EmailInfo {
        channel: Channel::Maildrop,
        email: email.clone(),
        token: Some(email),
        expires_at: None,
        created_at: None,
    })
}

/// 通过详情接口获取单封邮件完整内容
/// GET /api/email_content.php?id={id}
/// 详情响应字段（从前端代码确认）:
/// - content: 完整 HTML 正文
/// - subject / from_addr / date: 邮件元数据
/// - attachment: JSON 字符串数组 [{filename, path, size}]（可能为空）
async fn fetch_detail(id: &str) -> Option<Value> {
    let trimmed = id.trim();
    if trimmed.is_empty() {
        return None;
    }
    let url = format!(
        "{}/api/email_content.php?id={}",
        BASE,
        urlencoding::encode(trimmed)
    );
    let resp = md_headers(http_client().get(url)).send().await.ok()?;
    if !resp.status().is_success() {
        return None;
    }
    resp.json::<Value>().await.ok()
}

/// 解析详情接口的 attachment 字段（JSON 字符串）为附件列表
fn parse_attachments(raw: &Value) -> Value {
    if let Some(s) = raw.as_str() {
        if !s.trim().is_empty() {
            if let Ok(v) = serde_json::from_str::<Value>(s) {
                if v.is_array() {
                    return v;
                }
            }
        }
    }
    Value::Array(vec![])
}

pub fn get_emails(_token: &str, email: &str) -> Result<Vec<Email>, String> {
    let addr = email.trim();
    let addr = if addr.is_empty() { _token.trim() } else { addr };
    if addr.is_empty() {
        return Err("maildrop: empty address".into());
    }
    let q = urlencoding::encode(addr);
    let url = format!("{}/api/emails.php?addr={}&page=1&limit=20", BASE, q);
    block_on(async {
        let resp = md_headers(http_client().get(url))
            .send()
            .await
            .map_err(|e| e.to_string())?;
        if !resp.status().is_success() {
            return Err(format!("maildrop emails {}", resp.status()));
        }
        let data: Value = resp.json().await.map_err(|e| e.to_string())?;
        let rows = data
            .get("emails")
            .and_then(|x| x.as_array())
            .cloned()
            .unwrap_or_default();

        let mut out = Vec::with_capacity(rows.len());
        for raw in &rows {
            /* 提取 ID（支持 string/number） */
            let id = raw
                .get("id")
                .and_then(|v| {
                    v.as_str()
                        .map(|s| s.to_string())
                        .or_else(|| v.as_i64().map(|n| n.to_string()))
                        .or_else(|| v.as_u64().map(|n| n.to_string()))
                })
                .unwrap_or_default();

            /* 列表 description 作为文本回退 */
            let desc = raw
                .get("description")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .trim()
                .to_string();

            /* 拉取详情覆盖 html/from/subject/date/attachments */
            let mut from = raw
                .get("from_addr")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .trim()
                .to_string();
            let mut subject = raw
                .get("subject")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .trim()
                .to_string();
            let mut date = raw
                .get("date")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string();
            let mut html = String::new();
            let mut attachments = Value::Array(vec![]);

            if !id.is_empty() {
                if let Some(detail) = fetch_detail(&id).await {
                    if let Some(c) = detail.get("content").and_then(|v| v.as_str()) {
                        if !c.trim().is_empty() {
                            html = c.to_string();
                        }
                    }
                    if let Some(f) = detail.get("from_addr").and_then(|v| v.as_str()) {
                        if !f.trim().is_empty() {
                            from = f.trim().to_string();
                        }
                    }
                    if let Some(s) = detail.get("subject").and_then(|v| v.as_str()) {
                        if !s.trim().is_empty() {
                            subject = s.trim().to_string();
                        }
                    }
                    if let Some(d) = detail.get("date").and_then(|v| v.as_str()) {
                        if !d.trim().is_empty() {
                            date = d.to_string();
                        }
                    }
                    if let Some(att_raw) = detail.get("attachment") {
                        attachments = parse_attachments(att_raw);
                    }
                }
            }

            let is_read = raw
                .get("isRead")
                .and_then(|v| v.as_bool().or_else(|| v.as_i64().map(|n| n != 0)))
                .unwrap_or(false);

            let flat = serde_json::json!({
                "id": id,
                "from": from,
                "to": addr,
                "subject": subject,
                "text": desc,
                "html": html,
                "date": date,
                "isRead": is_read,
                "attachments": attachments,
            });
            out.push(normalize_email(&flat, addr));
        }
        Ok(out)
    })
}
